### ISSUE TITLE
Snowflake `identifier` - sqlfluff fix unable to parse

### ISSUE THREAD

**Issue by upjohnc** [sqlfluff/sqlfluff#3404]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When running `sqlfluff fix` on files that have a line with `identifier()` in it, the file cannot be parsed.  We have files that that use the `identifier` function in snowflake and we do a workaround of commenting out that line to then run `sqlfluff fix`.

This is an example of a line making use of the snowflake `identifier` function.

```sql
USE SCHEMA identifier('{{ params.schema_name }}');
```

### Expected Behaviour

When running `sqlfluff fix`, the command is able to change the file rather than returning a "parsing error".

### Observed Behaviour

when running `sqlfluff fix` on a file with the `identifier` function, the result is "parsing error"

```sql
USE SCHEMA identifier('{{ params.schema_name }}');
```

I can confirm that without the jinja it returns a parsing error

```sql
USE SCHEMA identifier('tt');
```

### How to reproduce

A file with a line like this, with `identifier`
```sql
USE SCHEMA identifier('{{ params.schema_name }}');
```
run sqlfluff fix on that file

### Dialect

snowflake

### Version

0.13.0

### Configuration

# Usage : README_SQLFULFF.md

[sqlfluff:core]
templater = jinja
dialect = snowflake
ignore = parsing
# L031 - weird table aliasing rule
# L032 - allow 'using', rather than "sqlfluff fix" changing to 'on' clause
# L034 - the order of elements in the select clause
# The reason is that it will change the order and effect the insert statement
# https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L034
exclude_rules = L009, L031, L032, L034

[sqlfluff:rules]
comma_style = trailing

[sqlfluff:rules:L016]
max_line_length = 300

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper

[sqlfluff:rules:L014]
# Keywords
extended_capitalisation_policy = lower

[sqlfluff:rules:L030]
# functions
extended_capitalisation_policy = upper

[sqlfluff:rules:L040]
# booleans and null
capitalisation_policy = upper

[sqlfluff:rules:L063]
# datatypes
extended_capitalisation_policy = upper

[sqlfluff:templater:jinja:context]
dag_run.conf = abc
params = {'schema_name': 'test', 'days_back': 1}
dag_run = {'conf': {'start_date': 'test', 'end_date': 'w', 'window_start_date': '2021-01-01'}}
ds = 'tt'

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by mdahlman** [comment]
Interesting.
`IDENTIFIER` is not a function in Snowflake. Here's their definitive list of all functions: https://docs.snowflake.com/en/sql-reference/functions-all.html

`IDENTIFIER` is a function in Snowflake. Here's the proof: https://docs.snowflake.com/en/sql-reference/identifier-literal.html

By the way, you mention `fix` in the issue summary. But really it's purely a shortcoming in `parse`. A minimal test case is something like this:
`echo "USE SCHEMA identifier('tt');" | sqlfluff parse --dialect snowflake -`

or this example straight from their docs:
`echo "create or replace table identifier('my_db.my_schema.my_table') (c1 number);" | sqlfluff parse --dialect snowflake -`

I suspect that the solution would be to override class `ObjectReferenceSegment` in the snowflake dialect to allow for the secret `IDENTIFIER` function.

**Comment by tunetheweb** [comment]
> IDENTIFIER is not a function in Snowflake.

The [docs note this](https://docs.snowflake.com/en/sql-reference/identifier-literal.html#usage-notes):
> Although `IDENTIFIER(...)` uses the syntax of a function, it is not a true function and is not returned by commands such as `SHOW FUNCTIONS`.

> I suspect that the solution would be to override class `ObjectReferenceSegment` in the snowflake dialect to allow for the secret `IDENTIFIER` function.

Sounds sensible to me.

**Comment by upjohnc** [comment]
That is good clarification on the nomenclature.  I guess I did assume that it was a function because of the syntax.  I am now curious what the name of the `identifier` object is.

But to the possible solution to the "issue".  When you say "override class `ObjectReferenceSegment`", what do you mean?  Honestly, I have only glanced through the structure of the sqlfluff code.  I figured I'd ask for clarification in case I can lend a hand.

**Comment by tunetheweb** [comment]
The `ObjectReferenceSegment` is the base class for all object references (table references, column references… etc.). It currently looks like this:

https://github.com/sqlfluff/sqlfluff/blob/0ba5457f07c6a55cebde82b9f4054401be6d6a2b/src/sqlfluff/dialects/dialect_ansi.py#L699-L708

We could override this in Snowflake and allow either that definition, or the `Identifier(..)` “function”, using the `OneOf` syntax, so that `Identifier(..)` would automatically be able to be used in an Object Reference. Or override `SingleIdentifierGrammar` for the same effect.

This would be much easier than replacing every ObjectReferenceSegment with a `OneOf` of this, which would be a much more involved change and not that scalable.

**Comment by mdahlman** [comment]
I'm with @upjohnc: It's all well and good to document the negative, "it is not a true function". But they should include the positive statement as well: "it is not a true function, rather it is an _xxx_". I referred to it as a pseudo-function in my code comments, but perhaps there's a more accurate/precise word.
